marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java

```diff
@@ -59,7 +62,59 @@ private BulkOutputCallerImpl(OutputEndpointImpl endpoint) {
         @Override
         public void awaitCompletion() {
             logger.trace("output endpoint running endpoint={} work={}", getEndpointPath(), getWorkUnit());
+            setPhase(WorkPhase.RUNNING);
+            SessionState session = allowsSession() ? getEndpoint().getCaller().newSessionState() : null;
 
+            calling: while (true) {
+                Stream<InputStream> output = null;
+                try {
+                    logger.trace("output endpoint={} count={} state={}",
+                            getEndpointPath(), getCallCount(), getEndpointState());
+
+                    output = getEndpoint().getCaller().call(
+                            getEndpoint().getClient(), getEndpointState(), session, getWorkUnit()
+                    );
+                    incrementCallCount();
+                } catch(Throwable throwable) {
+                    throw new RuntimeException("error while calling "+getEndpoint().getEndpointPath(), throwable);
+                }
+
+                if (allowsEndpointState()) {
```

Comment:
Should the client also process the output if there is no endpoint state?

Would it work to move the construction of the result array before this if branch and the for loop after the if branch.

The if branch would also need to increment i from zero to 1.